pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,178 @@
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KindCalicoNodeStatus     = "CalicoNodeStatus"
+	KindCalicoNodeStatusList = "CalicoNodeStatusList"
+)
+
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CalicoNodeStatusList is a list of CalicoNodeStatus resources.
+type CalicoNodeStatusList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []CalicoNodeStatus `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type CalicoNodeStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   CalicoNodeStatusSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status CalicoNodeStatusStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// CalicoNodeStatusSpec contains the specification for a CalicoNodeStatus resource.
+type CalicoNodeStatusSpec struct {
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
+	Classes []NodeStatusClassType `json:"classes,omitempty"`
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
+}
+
+// CalicoNodeStatusStatus defines the observed state of CalicoNodeStatus.
+// No validation needed for status since it is updated by Calico.
+type CalicoNodeStatusStatus struct {
+	// UpdateTimestamp is a timestamp representing the server time when CalicoNodeStatus object
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	UpdateTimestamp metav1.Time `json:"updateTimestamp,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
+
+	// BGPStatus holds node BGP status.
+	BGPStatus CalicoNodeBGPStatus `json:"bgpStatus,omitempty"`
+}
+
+// CalicoNodeBGPStatus defines the observed state of BGP status on the node.
+type CalicoNodeBGPStatus struct {
+	// The total number of established bgp sessions.
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
+
+	// BirdStatus represents the latest observed status of bird.
+	BirdStatus CalicoNodeBirdStatus `json:"birdStatus,omitempty"`
+
+	// Peers represents BGP peers status on the node.
+	Peers []CalicoNodePeers `json:"peers,omitempty"`
+
+	// Routes represents routes on the node.
+	Routes []CalicoNodeRoutes `json:"routes,omitempty"`
```

Comment:
I think Routes probably belongs outside of BGP?

We can learn routes from more than just BGP (e.g., from the kernel, or static routes, or from OSPF if we ever enable that).

We probably want each route object to have a field indicating from where it was learned.

Reply:
Agree. Let me move it out of BGP